File: cpp/2670.find-the-distinct-difference-array.cpp
#include <vector>
using std::vector;

class Solution {
public:
  vector<int> distinctDifferenceArray(vector<int> &nums) {
    vector<int> res;
    vector<int> prefix(51, 0), suffix(51, 0);

    int prefixCount = 0, suffixCount = 0;
    for (int num : nums) {
      suffix[num]++;
      if (suffix[num] == 1)
        suffixCount++;
    }

    for (int num : nums) {
      prefix[num]++;
      if (prefix[num] == 1)
        prefixCount++;

      suffix[num]--;
      if (suffix[num] == 0)
        suffixCount--;

      res.push_back(prefixCount - suffixCount);
    }

    return res;
  }
};

File: python/2670.find-the-distinct-difference-array.py
class Solution:
    def distinctDifferenceArray(self, nums: list[int]) -> list[int]:
        suffix: list[int] = [0] * 51
        prefix: list[int] = [0] * 51

        suffixCount = 0
        prefixCount = 0

        for num in nums:
            suffix[num] += 1
            if suffix[num] == 1:
                suffixCount += 1

        res: list[int] = []

        for num in nums:
            prefix[num] += 1
            suffix[num] -= 1

            if suffix[num] == 0:
                suffixCount -= 1
            if prefix[num] == 1:
                prefixCount += 1

            res.append(prefixCount - suffixCount)

        return res

